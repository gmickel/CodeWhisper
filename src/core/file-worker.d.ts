export function detectLanguage(filePath: string): string;
export function stripComments(code: string, language: string): string;
export default function processFile(options: {
  filePath: string;
  suppressComments: boolean;
}): Promise<object | null>;
