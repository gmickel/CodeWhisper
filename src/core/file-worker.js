/**
 * This file is configured to use a JavaScript worker file (`file-worker.js`)
 * because the `Piscina` worker-thread pool works better with a JavaScript file as the entry point.
 * This ensures compatibility in both development (`src`), testing (`tests`), and production (`dist`).
 */
import fs from 'fs-extra';
import { isBinaryFile } from 'isbinaryfile';
import { stripComments } from '../utils/comment-stripper';
import { detectLanguage } from '../utils/language-detector';

/**
 * Process a file and return its metadata and content.
 *
 * @param {Object} options - The options for processing the file.
 * @param {string} options.filePath - The path of the file to process.
 * @param {boolean} options.suppressComments - Whether to suppress comments in the file content.
 * @returns {Promise<Object|null>} A promise that resolves to an object containing the file metadata and content,
 *                                or null if there was an error processing the file.
 */
export default async function processFile({ filePath, suppressComments }) {
  try {
    const stats = await fs.stat(filePath);
    const buffer = await fs.readFile(filePath);

    if (await isBinaryFile(buffer)) {
      return null;
    }

    let content = buffer.toString('utf-8');
    const extension = filePath.split('.').pop() ?? '';
    const language = detectLanguage(filePath);

    if (suppressComments) {
      content = stripComments(content, language);
    }

    return {
      path: filePath,
      extension,
      language,
      size: stats.size,
      created: stats.birthtime,
      modified: stats.mtime,
      content,
    };
  } catch (error) {
    console.error(`Error processing file ${filePath}:`, error);
    return null;
  }
}
