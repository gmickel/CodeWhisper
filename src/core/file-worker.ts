import fs from 'fs-extra';
import { isBinaryFile } from 'isbinaryfile';
import { stripComments } from '../utils/comment-stripper';
import { detectLanguage } from '../utils/language-detector';

interface WorkerData {
  filePath: string;
  suppressComments: boolean;
}

export default async function processFile({
  filePath,
  suppressComments,
}: WorkerData) {
  try {
    const stats = await fs.stat(filePath);
    const buffer = await fs.readFile(filePath);

    if (await isBinaryFile(buffer)) {
      return null;
    }

    let content = buffer.toString('utf-8');
    const extension = filePath.split('.').pop() ?? '';
    const language = detectLanguage(filePath);

    if (suppressComments) {
      content = stripComments(content, language);
    }

    return {
      path: filePath,
      extension,
      language,
      size: stats.size,
      created: stats.birthtime,
      modified: stats.mtime,
      content,
    };
  } catch (error) {
    console.error(`Error processing file ${filePath}:`, error);
    return null;
  }
}
