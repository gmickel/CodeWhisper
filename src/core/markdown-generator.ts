import path from 'node:path';
import Handlebars from 'handlebars';
import type { FileInfo } from './file-processor';

export interface MarkdownOptions {
  template?: string;
  noCodeblock?: boolean;
  customData?: Record<string, unknown>;
  basePath?: string;
}

function registerHandlebarsHelpers(noCodeblock: boolean) {
  Handlebars.registerHelper(
    'codeblock',
    (content: string, language: string) => {
      if (noCodeblock) {
        return content;
      }
      return new Handlebars.SafeString(`\`\`\`${language}\n${content}\n\`\`\``);
    },
  );

  Handlebars.registerHelper('eq', (v1, v2) => v1 === v2);
  Handlebars.registerHelper('objectKeys', Object.keys);
  Handlebars.registerHelper('gt', (a, b) => a > b);

  Handlebars.registerHelper('hasCustomData', (context) =>
    Object.keys(context).some(
      (key) => key !== 'files' && key !== 'base' && Object.hasOwn(context, key),
    ),
  );

  Handlebars.registerHelper(
    'isCustomData',
    (key) => key !== 'files' && key !== 'base',
  );

  Handlebars.registerHelper('lineNumbers', (content: string) => {
    const lines = content.split('\n');
    const numberedLines = lines.map((line, index) => `${index + 1} | ${line}`);
    return new Handlebars.SafeString(numberedLines.join('\n'));
  });

  Handlebars.registerHelper('tableOfContents', (files: FileInfo[], options) => {
    const basePath = options.data.root.base;
    const toc = files
      .map((file) => `- ${relativePath(basePath, file.path)}`)
      .join('\n');
    return new Handlebars.SafeString(toc);
  });

  Handlebars.registerHelper('fileInfo', (file: FileInfo, options) => {
    const basePath = options.data.root.base;
    return new Handlebars.SafeString(`
- Path: ${relativePath(basePath, file.path)}
- Language: ${file.language}
- Size: ${file.size} bytes
- Last modified: ${file.modified}
`);
  });

  Handlebars.registerHelper('relativePath', (filePath: string, options) => {
    const basePath = options.data.root.base;
    return relativePath(basePath, filePath);
  });

  // Adapt to handle missing helpers gracefully
  Handlebars.registerHelper('helperMissing', (...args) => {
    const options = args.pop();
    console.warn(`Missing helper: "${options.name}"`);
    return `Missing helper: "${options.name}"`;
  });
}

function relativePath(base: string, target: string): string {
  return path.relative(base, target).replace(/\\/g, '/');
}

export async function generateMarkdown(
  files: FileInfo[],
  templateContent: string,
  options: MarkdownOptions = {},
): Promise<string> {
  const {
    noCodeblock = false,
    customData = {},
    basePath = process.cwd(),
  } = options;

  registerHandlebarsHelpers(noCodeblock);

  const compiledTemplate = Handlebars.compile(templateContent);

  const data = {
    files,
    base: basePath,
    ...customData,
  };

  const result = compiledTemplate(data);

  // Normalize line endings to LF
  return result.replace(/\r\n/g, '\n');
}
