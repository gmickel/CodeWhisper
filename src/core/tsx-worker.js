import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { Worker } from 'node:worker_threads';

// Resolve the path to tsx/cli for executing TypeScript files
const tsx = fileURLToPath(import.meta.resolve('tsx/cli'));
const __dirname = dirname(fileURLToPath(import.meta.url));

// Worker options (Replace 'file-worker.ts' with the path to your worker TypeScript file)
const workerFile = resolve(__dirname, 'file-worker.ts');

// Create a worker using tsx to handle TypeScript execution
const worker = new Worker(tsx, { argv: [workerFile] });

export default worker;
