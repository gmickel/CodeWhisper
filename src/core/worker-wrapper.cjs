const { workerData } = require('node:worker_threads');

if (workerData.fullpath.endsWith('.ts')) {
  require('ts-node').register();
}
module.exports = require(workerData.fullpath);
