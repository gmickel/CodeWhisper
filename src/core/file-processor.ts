import os from 'node:os';
import path, { resolve } from 'node:path';
import fastGlob from 'fast-glob';
import fs from 'fs-extra';
import ignore from 'ignore';
import { isBinaryFile } from 'isbinaryfile';
import Piscina from 'piscina';
import { FileCache } from '../utils/file-cache';
import { filename } from './file-worker';

export interface FileInfo {
  path: string;
  extension: string;
  language: string;
  size: number;
  created: Date;
  modified: Date;
  content: string;
}

interface ProcessOptions {
  path: string;
  gitignorePath?: string;
  filter?: string[];
  exclude?: string[];
  suppressComments?: boolean;
  caseSensitive?: boolean;
  customIgnores?: string[];
  cachePath?: string;
}

const isDist = path
  .dirname(new URL(import.meta.url).pathname)
  .includes('/dist/');
const workerFilename = 'file-worker.js';
const workerFilePath = path.resolve(
  path.dirname(new URL(import.meta.url).pathname),
  '../core',
  workerFilename,
);

const pool = new Piscina({ filename: workerFilePath });

const DEFAULT_IGNORES = [
  // Version control
  '**/.git',
  '**/.gitignore',
  '**/.gitattributes',
  '**/.gitmodules',
  '**/.gitkeep',
  '**/.github',
  '**/.svn',
  '**/.hg',
  '**/.hgignore',
  '**/.hgcheck',

  // Package manager directories
  '**/node_modules',
  '**/jspm_packages',
  '**/bower_components',

  // Build outputs and caches
  '**/dist',
  '**/build',
  '**/.cache',
  '**/.output',
  '**/.nuxt',
  '**/.next',
  '**/.vuepress/dist',
  '**/.serverless',
  '**/.fusebox',
  '**/.dynamodb',

  // Package manager files
  '**/package-lock.json',
  '**/yarn.lock',
  '**/pnpm-lock.yaml',
  '**/bun.lockb',
  '**/Gemfile.lock',
  '**/Cargo.lock',
  '**/poetry.lock',
  '**/composer.lock',

  // IDE and editor files
  '**/.vscode',
  '**/.idea',
  '**/*.swp',
  '**/*.swo',

  // OS generated files
  '**/.DS_Store',
  '**/.DS_Store?',
  '**/._*',
  '**/.Spotlight-V100',
  '**/.Trashes',
  '**/ehthumbs.db',
  '**/Thumbs.db',

  // Logs
  '**/logs',
  '**/*.log',
  '**/npm-debug.log*',
  '**/yarn-debug.log*',
  '**/yarn-error.log*',

  // Temporary files
  '**/tmp',
  '**/temp',
];

export async function processFiles(
  options: ProcessOptions,
): Promise<FileInfo[]> {
  const {
    path: basePath,
    gitignorePath = '.gitignore',
    filter = [],
    exclude = [],
    suppressComments = false,
    caseSensitive = false,
    customIgnores = [],
    cachePath = path.join(os.tmpdir(), 'codewhisper-cache.json'),
  } = options;

  const fileCache = new FileCache(cachePath);

  const ig = ignore().add(DEFAULT_IGNORES).add(customIgnores);

  if (await fs.pathExists(gitignorePath)) {
    const gitignoreContent = await fs.readFile(gitignorePath, 'utf-8');
    ig.add(gitignoreContent);
  }

  const globOptions: fastGlob.Options = {
    cwd: basePath,
    dot: true,
    absolute: true,
    onlyFiles: true,
    followSymbolicLinks: false,
    caseSensitiveMatch: caseSensitive,
    ignore: exclude,
  };

  const fileInfos: FileInfo[] = [];
  const tasks: Promise<void>[] = [];

  const globStream = fastGlob.stream('**/*', globOptions);

  return new Promise((resolve, reject) => {
    globStream.on('data', (filePath) => {
      tasks.push(
        (async () => {
          const filePathStr = filePath.toString();
          const relativePath = path.relative(basePath, filePathStr);

          if (ig.ignores(relativePath)) return;

          if (
            filter.length > 0 &&
            !filter.some((pattern) => new RegExp(pattern).test(filePathStr))
          )
            return;

          try {
            const cached = await fileCache.get(filePathStr);
            if (cached) {
              fileInfos.push(cached);
              return;
            }

            const stats = await fs.stat(filePathStr);
            if (!stats.isFile()) return;

            const buffer = await fs.readFile(filePathStr);
            if (await isBinaryFile(buffer)) return;

            const result = await pool.run({
              filePath: filePathStr,
              suppressComments,
            });

            if (result) {
              await fileCache.set(filePathStr, result as FileInfo);
              fileInfos.push(result as FileInfo);
            }
          } catch (error) {
            console.error(`Error processing file ${filePathStr}:`, error);
          }
        })(),
      );
    });

    globStream.on('end', async () => {
      await Promise.all(tasks);
      fileInfos.sort((a, b) => a.path.localeCompare(b.path));
      resolve(fileInfos);
    });

    globStream.on('error', reject);
  });
}

export const testExports = {
  pool,
};
