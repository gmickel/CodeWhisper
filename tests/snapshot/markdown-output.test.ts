import path from 'node:path';
import fs from 'fs-extra';
import { describe, expect, it } from 'vitest';
import { type FileInfo, processFiles } from '../../src/core/file-processor';
import { generateMarkdown } from '../../src/core/markdown-generator';

const joinPath = (pathname: string) =>
  new URL(pathname, import.meta.url).pathname;

describe('Markdown Output', () => {
  const fixturesPath = joinPath('../fixtures');
  const customTemplatePath = path.join(fixturesPath, 'custom-template.hbs');
  const defaultTemplatePath = path.join(fixturesPath, 'default.hbs');
  const testDir = path.join(fixturesPath, 'test-project');

  async function readTemplateContent(templatePath: string): Promise<string> {
    return await fs.readFile(templatePath, 'utf-8');
  }

  async function getProcessedFiles(directory: string) {
    return await processFiles({ path: directory });
  }

  async function generateMarkdownContent(
    files: FileInfo[],
    templateContent: string,
    options: Record<string, unknown>,
  ) {
    return await generateMarkdown(files, templateContent, {
      ...options,
      basePath: (options.basePath as string) || testDir,
    });
  }

  it('should match snapshot for default template', async () => {
    const files = await getProcessedFiles(testDir);
    const defaultTemplateContent =
      await readTemplateContent(defaultTemplatePath);

    const markdown = await generateMarkdownContent(
      files,
      defaultTemplateContent,
      {},
    );

    expect(markdown).toMatchSnapshot();
  });

  it('should match snapshot for custom template with custom data', async () => {
    const files = await getProcessedFiles(testDir);
    const customTemplateContent = await readTemplateContent(customTemplatePath);

    const markdownWithCustomData = await generateMarkdownContent(
      files,
      customTemplateContent,
      {
        customData: { title: 'Snapshot Test Project', version: '1.0.0' },
      },
    );

    expect(markdownWithCustomData).toMatchSnapshot();
  });

  it('should match snapshot for default template with additional custom data', async () => {
    const files = await getProcessedFiles(testDir);
    const defaultTemplateContent =
      await readTemplateContent(defaultTemplatePath);

    const markdownWithCustomData = await generateMarkdownContent(
      files,
      defaultTemplateContent,
      {
        customData: { key: 'value', another: 'data point' },
      },
    );

    expect(markdownWithCustomData).toMatchSnapshot();
  });
});
