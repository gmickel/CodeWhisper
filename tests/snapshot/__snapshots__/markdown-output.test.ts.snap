// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Markdown Output > should match snapshot for custom template with custom data 1`] = `
"# Custom Template: Snapshot Test Project

## Project Overview

This project contains 0 file(s).

## File Listing


## Custom Data

Custom data provided:
- title: Snapshot Test Project
- version: 1.0.0

## Generated on: Missing helper: &quot;now&quot;
"
`;

exports[`Markdown Output > should match snapshot for default template 1`] = `
"# Code Summary

- /Users/gordon/work/CodeWhisper/tests/fixtures/test-project/package.json
- /Users/gordon/work/CodeWhisper/tests/fixtures/test-project/src/utils.ts
- /Users/gordon/work/CodeWhisper/tests/fixtures/test-project/src/main.js

## Files

## /Users/gordon/work/CodeWhisper/tests/fixtures/test-project/package.json

- Language: json
- Size: 51 bytes
- Last modified: Tue Jul 09 2024 22:40:41 GMT+0200 (Central European Summer Time)

\`\`\`json
{
  "name": "test-project",
  "version": "1.0.0"
}

\`\`\`

## /Users/gordon/work/CodeWhisper/tests/fixtures/test-project/src/utils.ts

- Language: typescript
- Size: 70 bytes
- Last modified: Tue Jul 09 2024 22:40:30 GMT+0200 (Central European Summer Time)

\`\`\`typescript
export function add(a: number, b: number): number {
  return a + b;
}

\`\`\`

## /Users/gordon/work/CodeWhisper/tests/fixtures/test-project/src/main.js

- Language: javascript
- Size: 30 bytes
- Last modified: Tue Jul 09 2024 22:40:24 GMT+0200 (Central European Summer Time)

\`\`\`javascript
console.log('Hello, World!');

\`\`\`

"
`;

exports[`Markdown Output > should match snapshot for default template with additional custom data 1`] = `
"# Code Summary



## Files

"
`;
